### PR TITLE
Add new check for indexing with atomic=True

### PR DESCRIPTION
**Pull request by DavidCain**
Briefly, this check helps prevent a long-running index creation from
preventing all reads and writes to a table.

This supplements an existing check
==================================
There is an existing check, `CREATE_INDEX` which warns about creating
indices without the `CONCURRENTLY` argument. It's a helpful check!

However, there are some valid reasons to prefer non-concurrent index
creation. For example, concurrent index creation may struggle to
complete when applied to write-heavy tables (since Postgres will have to
make repeated passes on the table to accommodate the new reads since the
last round).

The danger of `EXCLUSIVE` locks
===============================
If you *do* want to create indices nonconcurrently, there is still a
**huge footgun** to be aware of -- locks which are held in a transaction
are not released until the transaction closes.

Dangerous Django defaults
=========================
By default, any time you add a new Django field with `db_index=True` or
`unique=True`, the automatically generated `Migration` has:

- implicit setting of `atomic=True` (transactions are on)
- an `ALTER TABLE...ADD COLUMN` call (exclusively locks the table)
- a `CREATE INDEX` invocation (*hopefully* quick, due to nullable
  columns)

It's even more dangerous if you combine an `ALTER TABLE` call with
`CREATE INDEX` of an existing, non-nullable column.

Favor `atomic=False` if you must index
======================================
Index-creating migrations should have few `operations`. Since index
creation does not modify row-level data (and only requires locking
against writes), there's no reason to use transactions.

It's prudent to bypass the default `atomic=True` for index creation.

### CODE REVIEW

**Comment by DavidCain** [comment]
@David-Wobrock - is this repository still maintained and/or open to contributions? Do you have thoughts about this proposed check?

File: django_migration_linter/sql_analyser/postgresql.py [hunk 1]
```diff
@@ -5,6 +5,36 @@
...
+    """
+    if not (sql_statements and sql_statements[0].startswith("BEGIN")):
+        return False
+
+    for i, sql in enumerate(sql_statements):
+        # If any statements acquire an exclusive lock, complain about index creation later.
+        # Nearly every single `ALTER TABLE` command requires an exclusive lock:
+        #     https://www.postgresql.org/docs/current/sql-altertable.html
+        # (Most common example is `ALTER TABLE... ADD COLUMN`, then later `CREATE INDEX`)
+        if sql.startswith("ALTER TABLE"):
```
Comment (fevral13) [inline django_migration_linter/sql_analyser/postgresql.py:33]: This check may give false positives for operations, that do not take long time, like dropping columns, indexes or anything else. I suggest to explicitly specify the list of potentially long operations.
```suggestion
        long_operations = ("ADD COLUMN", "CREATE INDEX")
        column_or_index_creation = any(op in sql for op in long_operations)
        if sql.startswith("ALTER TABLE") and column_or_index_creation:
```

Comment (DavidCain) [reply django_migration_linter/sql_analyser/postgresql.py:33]: As an example, consider the following SQL:
```sql
BEGIN;
ALTER TABLE "users" DROP COLUMN "last_name";  -- fast, EXCLUSIVE
CREATE INDEX "user_email" ON "users" ("email");  -- slow, SHARED
COMMIT;
```

Even though the `DROP COLUMN` should be very fast, the exclusive lock it requires is held until the indexing operation releases.

The problem goes away if you either:
- stop using transactions
- remove the `ALTER TABLE` that comes before index creation

**Comment by DavidCain** [comment]
Thanks for the review, @fevral13 ! I tried to leave some comments explaining the counterintuitive nature of why even a fast `ALTER TABLE` operation is problematic when it comes *before* index creation.

File: django_migration_linter/sql_analyser/postgresql.py (hunk 1 above)
Comment (fevral13) [reply django_migration_linter/sql_analyser/postgresql.py:33]: Ok, David, thank you for explanation. This is one of the things I did not know about Postgesql. I am ok with this check. What however I find missing, is more informative error message, and the most important - ways to resolve it. Please, improve the error message, so that it not only explains the problem, but also how to fix it. Something like: you need to either disable transaction by adding xxx, or split this migration in 2 parts to separate expensive operations.

**Comment by David-Wobrock** [comment]
Thanks for the PR! I'll try to have a look in the next days 😇

I need to find some time to give some love to this repo again! Many requests and improvements that can be made 💪

File: django_migration_linter/sql_analyser/postgresql.py (hunk 1 above)
Comment (DavidCain) [reply django_migration_linter/sql_analyser/postgresql.py:33]: That's an excellent suggestion, and I'll make efforts to improve the clarity here. Thank you for your feedback!

File: django_migration_linter/sql_analyser/postgresql.py [hunk 2]
```diff
@@ -5,6 +5,36 @@
...
+            return has_create_index(sql_statements[i + 1 :])
```
Comment (David-Wobrock) [inline django_migration_linter/sql_analyser/postgresql.py:34]: The issue re-using this function here, is that we are not detecting some cases you are describing. For instance if we had this unit test:
```
diff --git a/tests/unit/test_sql_analyser.py b/tests/unit/test_sql_analyser.py
index 01a2bb3..7553796 100644
--- a/tests/unit/test_sql_analyser.py
+++ b/tests/unit/test_sql_analyser.py
@@ -320,6 +320,15 @@ class PostgresqlAnalyserTestCase(SqlAnalyserTestCase):
         ]
         self.assertBackwardIncompatibleSql(sql, code="CREATE_INDEX_EXCLUSIVE")

+    def test_create_concurrently_index_exclusive(self):
+        sql = [
+            "BEGIN;",
+            'ALTER TABLE "users" ADD COLUMN "email" varchar(254) NULL;',
+            'CREATE CONCURRENTLY INDEX "user_email" ON "users" ("email");',
+            "COMMIT;",
+        ]
+        self.assertBackwardIncompatibleSql(sql, code="CREATE_INDEX_EXCLUSIVE")
+
     def test_create_index_exclusive_no_lock(self):
         sql = [
             'ALTER TABLE "users" ADD COLUMN "email" varchar(254) NULL;',
```
It does not trigger an error:
```
Found 19 test(s).
System check identified no issues (0 silenced).
..F................
======================================================================
FAIL: test_create_concurrently_index_exclusive (tests.unit.test_sql_analyser.PostgresqlAnalyserTestCase.test_create_concurrently_index_exclusive)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./django-migration-linter/tests/unit/test_sql_analyser.py", line 330, in test_create_concurrently_index_exclusive
    self.assertBackwardIncompatibleSql(sql, code="CREATE_INDEX_EXCLUSIVE")
  File "./django-migration-linter/tests/unit/test_sql_analyser.py", line 30, in assertBackwardIncompatibleSql
    self.assertNotEqual(0, len(errors), "Found no errors in sql")
AssertionError: 0 == 0 : Found no errors in sql

----------------------------------------------------------------------
Ran 19 tests in 0.003s

FAILED (failures=1)
```

I guess we can modify a bit the check to also take that into account :) Since, as you said, creating an index concurrently can be a long running operation.

File: django_migration_linter/sql_analyser/postgresql.py [hunk 3]
```diff
@@ -32,6 +62,13 @@ class PostgresqlAnalyser(BaseAnalyser):
             mode=CheckMode.TRANSACTION,
             type=CheckType.WARNING,
         ),
+        Check(
+            code="CREATE_INDEX_EXCLUSIVE",
+            fn=has_create_index_in_transaction,
+            message="CREATE INDEX prolongs transaction, delaying lock release",
+            mode=CheckMode.TRANSACTION,
+            type=CheckType.ERROR,
```
Comment (David-Wobrock) [inline django_migration_linter/sql_analyser/postgresql.py:70]: I think this is the main point I see with this contribution.

The check makes total sense, and thanks for adding it 👍 We definitely want to merge it :)

1/ It's not a silver bullet. I don't think by default we should raise an error during migration linting when doing this. To me, we should warn the user instead, that it could create a long-running locking transaction. Basically, stay in the same vein as the other indexing operations which also lock the table. It's the same risk => we know it will lock the table, but we can only warn that it is potentially a long-running lock, since we don't know about the real size and load of the table.

In short:
```suggestion
            type=CheckType.WARNING,
```
😁 

2/ It feels like a subset of `CREATE_INDEX`, since it checks `CREATE_INDEX` + being in a transaction + having an `ALTER TABLE` operation. I feel like there is room for something more generic, and not only when linked to index creation.
Basically, any combination of "transaction + exclusive lock operation + potential long-running operation" should trigger such a warning.
I'm completely fine starting with this specific check, we don't have to change the approach, but it's more to keep in mind that we could enhance this check to something more generic 🙂

**Review by David-Wobrock — CHANGES_REQUESTED** [review]
Great contribution @DavidCain Thanks a lot! 💯 

I have two little comments to address: handle `concurrently` index creation, and change the check to a warning. With these changes, we should be good to merge it IMO 👍

File: django_migration_linter/sql_analyser/postgresql.py (hunk 3 above)
Comment (DavidCain) [reply django_migration_linter/sql_analyser/postgresql.py:70]: 1. You're absolutely right to point out that this *could* have some false positives (i.e. indexing a small table will generally be fast & thus there's little cause for concern). I opted for an `ERROR` because I can't really think of many situations in which you would ever *need* to be creating an index within a transaction. I understand your rationale, though -- we should probably only call it an error if it will _always_ be the source of problems.

   I'll happily make this a `WARNING` and reach for `--warnings-as-errors` to elevate the warning for our use cases.

2. Also totally agreed that this could be made generic for detecting _any_ long-running queries that prolong lock release. I think the reason I targeted `CREATE INDEX` is that:
    - index creation is a very common migration operation
    - you can pretty much *always* move index creation out of the transaction. Other schema-modifying queries may not share that easy resolution.

   I would also be very happy to _start_ with this type of check, and replace it entirely with something more generic at a later time!
   
Thanks for the review & the thorough explanations! I'll be applying both your suggested changes, I think they'll help a ton.

File: django_migration_linter/sql_analyser/postgresql.py (hunk 2 above)
Comment (DavidCain) [reply django_migration_linter/sql_analyser/postgresql.py:34]: Great callout and this is just oversight on my part. I'll make sure that `CREATE INDEX CONCURRENTLY` is flagged as well.

File: django_migration_linter/sql_analyser/postgresql.py (hunk 3 above)
Comment (David-Wobrock) [reply django_migration_linter/sql_analyser/postgresql.py:70]: > I'll happily make this a WARNING and reach for --warnings-as-errors to elevate the warning for our use cases.

That would be great! Thanks a lot 🙇 
You can always use `--warnings-as-errors CREATE_INDEX_EXCLUSIVE` to mark only this one warning as an error. I think that should work 🤞 

> I would also be very happy to start with this type of check, and replace it entirely with something more generic at a later time!

Entirely agreed 👍  It was just an opening for the future, I'm completely fine with the current version, handling `CREATE INDEX [CONCURRENTLY]`.

**Comment by David-Wobrock** [comment]
An additional nice touch would be to add a mention of this change in the `CHANGELOG.md` file, as a new feature. 😇 

Once you did all the changes, we'll be able to merge this PR - and we'll release 5.0.0 of the linter 😄 It's been a long time since the last release.

**Comment by DavidCain** [comment]
Thanks again for the feedback! I _believe_ the outstanding requests have been addressed (it's a `WARNING` now, I test `CONCURRENTLY`, and I added a quick note to the `CHANGELOG`.

Thanks again for a really pleasant first PR experience. 😄 I'd be delighted to make any other changes you as see fit.

**Review by David-Wobrock — APPROVED** [review]
Great work @DavidCain 💯 
Thanks a lot for the contribution 🙇
